### PR TITLE
Ordering for _netdev devices

### PR DESCRIPTION
**Pull request by Werkov**
Please see respective commit messages for explanation of the changes.

TL;DR Network devices (e.g. iSCSI) timeouts are effectively (and confusingly) reduced by network initialization time.

To demonstrate the effect of the patch a dummy device unit with 30 seconds job timeout was specified. Compare the ordering against the `network-online.target`.

Without patches:
> Jan 26 18:56:04 localhost systemd[1]: mnt-dummy.mount: Installed new job mnt-dummy.mount/start as 263
> Jan 26 18:56:04 localhost systemd[1]: dev-dummy\x2ddevice.device: Installed new job dev-dummy\x2ddevice.device/start as 264
> Jan 26 18:56:31 germ79 systemd[1]: Reached target Network.
> Jan 26 18:56:31 germ79 systemd[1]: Reached target Network is Online.
> Jan 26 18:56:35 germ79 systemd[1]: dev-dummy\x2ddevice.device: Job dev-dummy\x2ddevice.device/start timed out.
> Jan 26 18:56:35 germ79 systemd[1]: dev-dummy\x2ddevice.device: Job dev-dummy\x2ddevice.device/start finished, result=timeout

With the patches:
> Jan 26 19:11:52 localhost systemd[1]: dev-dummy\x2ddevice.device: Installed new job dev-dummy\x2ddevice.device/start as 265
> Jan 26 19:11:52 localhost systemd[1]: mnt-dummy.mount: Installed new job mnt-dummy.mount/start as 264
> Jan 26 19:12:17 germ79 systemd[1]: Reached target Network.
> Jan 26 19:12:17 germ79 systemd[1]: Reached target Network is Online.
> Jan 26 19:12:47 germ79 systemd[1]: dev-dummy\x2ddevice.device: Job dev-dummy\x2ddevice.device/start timed out.
> Jan 26 19:12:47 germ79 systemd[1]: dev-dummy\x2ddevice.device: Job dev-dummy\x2ddevice.device/start finished, result=timeout

### CODE REVIEW

**Comment by fsateler** [comment]
This fails to build: `../src/shared/generator.c:232:37: error: format ‘%s’ expects a matching ‘char *’ argument [-Werror=format=]`

**Comment by Werkov** [comment]
Oh, sorry, that was a stale version. I've pushed the correct one now.

File: src/shared/generator.c [hunk 1]
```diff
@@ -192,6 +192,46 @@ int generator_write_timeouts(
...
+
+        if (!fstab_test_option(opts, "_netdev\0"))
+                return 0;
+
+        node = fstab_node_to_udev_node(what);
+        if (!node)
+                return log_oom();
+
+        if (!is_device_path(node)) {
+                log_info("Not applying _netdev option to non-device %s.", what);
```
Comment (keszybz) [inline src/shared/generator.c]: This should be (at most) log_debug. In fact I think you should just remove the log entry.

**Comment by poettering** [comment]
I don't like the JobTimeoutSec= patch I must say. First of all it's explicitly documented that the timeout applies to the job being queued. But secondly, this would break concepts such as this: 58f2fab16da947052756b7f9ace40f6ee7fa1519

**Comment by poettering** [comment]
So, my original thinking with the JobTimeoutSec= setting is that it is an overall timeout, and that per-unit timeouts are available in StartTimeoutSec= anyway. However, that doesn't really apply to device units, as they are externally managed, and systemd doesn't define a startup phase for them...

I do understand your usecase however. I wonder though if it wouldn't be nicer to simply increase the job timeout appropriately to deal with the extra time a dependency might take?

It would be an option to add a second timeout setting for all of this... JobRunningTimeoutSec= or so, but before we add that I think it would be great if we could see if we can do without it...

File: src/shared/generator.c (hunk 1 above)
Comment (Werkov) [reply src/shared/generator.c]: I force-pushed a change of this. On second thought, the information is of no use in runtime.

**Comment by Werkov** [comment]
I see the distinction between JobTimeoutSec= being sort of an envelope and {Start,Stop}TimeoutSec= timeout for the actual operation. Since device units are asynchronous we only have the envelope to work with, OTOH, we may make assumptions about the moment when timer should start ticking.

Thanks for pointing out a non-device unit with non-infinite job timeout. We cannot thus change the JobTimeoutSec= concept.

Increased timeout is IMO the simplest solution, I'm not convinced it's also the nicest. Drawback of such approach is that it's a workaround that one has to bear in mind when configuring network devices despite it can be handled automatically (with _netdev flag being enough) in more systematic way.

JobRunningTimeoutSec= would fit the idea for devices, however, it may be superfluous for other types of units. What about ExpectingTimeoutSec= that'd be specific to device units only?

**Comment by poettering** [comment]
> What about ExpectingTimeoutSec= that'd be specific to device units only?

Hmpf, the job engine doesn't know the differences between the unit types... for the job engine all unit types are the same. such a timeout however would not have any influence on the unit state engine, only on the job state engine, hence it would appear strange to me to expose it as a device unit type property...

But yeah, this is all a bit ugly... 

but maybe we can actually sell JobRunningTimeoutSec= as a good thing even for service and other units. That's because StartTimeoutSec= for service units applies individually on all commands listed as ExecStart= and the suchlike ExecXYZ= settings. A service which has 3x ExecStartPre= followed by a ExecStart= the StartTimeoutSec= is hence applied 4x individually. With the JobRunningTimeoutSec= we could define a timeout that applies to the combination of the four...

so maybe not that ugly after all?

**Comment by Werkov** [comment]
> StartTimeoutSec= for service units applies individually on all commands listed

Nice, another use case for running job timeout.
It shouldn't be that complicated to update the patch accordingly...

**Comment by mwilck** [comment]
@poettering: thanks for your thoughts.

On your suggestion to simply increase timeouts: That was our first idea. The point is that it's counter-intuitive, in particular for things like _netdev mounts. Our customer was arguing that it was an outright bug that a _netdev mount would time out 10s after (or even before!) the network came up, and I couldn't figure good arguments against that.

About [58f2fab](https://github.com/systemd/systemd/commit/58f2fab16da947052756b7f9ace40f6ee7fa1519): I noticed that, but according to the tests I did the semantic change for JobTimeoutSec=  did **not** break it. I tested with a dead NFS mount, the typical reason for shutdown waiting forever, and the timeout was triggered correctly. I suppose that at least in my case, the time difference between the shutdown job being set up and switching to running state was negligible. Of course I can't prove that no other shutdown scenarios or other usecases for JobTimeoutSec may be broken.

I also experimented with changes to the implementation of device units, such as implementing a fake "start" method for them, but IMO that brings about even worse semantic changes, and it IIRC doesn't even work reliably.

This is really something that belongs into  the generic "Job" concept. So, if changing the semantics of JobTimeoutSec really breaks stuff or is otherwise unacceptable, we need to go for JobRunningTimeoutSec=. But before doing that, I wished to have exactly this discussion.

**Comment by Werkov** [comment]
At last, I got down to do some tests with `JobRunningTimeoutSec=`. I've force-pushed the changes into the PR branch. Semantics of `JobTimeoutSec=` is kept and the new directive is implicitely used for device units.

File: src/core/load-fragment-gperf.gperf.m4 [hunk 2]
```diff
@@ -194,6 +194,7 @@ Unit.OnFailureIsolate,           config_parse_job_mode_isolate,      0,
 Unit.IgnoreOnIsolate,            config_parse_bool,                  0,                             offsetof(Unit, ignore_on_isolate)
 Unit.IgnoreOnSnapshot,           config_parse_warn_compat,           DISABLED_LEGACY,               0
 Unit.JobTimeoutSec,              config_parse_sec_fix_0,             0,                             offsetof(Unit, job_timeout)
+Unit.JobRunningTimeoutSec,       config_parse_sec_fix_0,             0,                             offsetof(Unit, job_running_timeout)
```
Comment (poettering) [inline src/core/load-fragment-gperf.gperf.m4]: please don't use "fix_0" here. That's just a hacked version we keep around for compat with old systemd versions

(in old systemd versions we permitted setting timeouts to "0" for disabling them, but since normalized that so that only a timeout of infinity turns off timeouts. For new settings we don't need this sort of compat however, hence please use normal config_parse_sec() here)

File: src/core/unit.c [hunk 3]
```diff
@@ -1336,6 +1337,10 @@ int unit_load(Unit *u) {
                         goto fail;
                 }
 
+                if (u->job_running_timeout != USEC_INFINITY && u->job_running_timeout > u->job_timeout) {
+                        log_unit_warning(u, "JobRunningTimeout is greater than JobTimeout, the latter will override.");
+                }
```
Comment (poettering) [inline src/core/unit.c]: no {} around single-line if blocks please, see CODING_STYLE

Comment (poettering) [reply src/core/unit.c]: also, in all documentation and log messages we suffix config file settings with "=" to indicate that they are config file settings. hence, please write this as "JobRunningTimeout= is greater than JobTimeout="...

File: src/shared/generator.c [hunk 4]
```diff
@@ -188,7 +188,7 @@ int generator_write_timeouts(
 
         return write_drop_in_format(dir, unit, 50, "device-timeout",
                                     "# Automatically generated by %s\n\n"
-                                    "[Unit]\nJobTimeoutSec=%s",
+                                    "[Unit]\nJobRunningTimeoutSec=%s",
                                     program_invocation_short_name, timeout);
```
Comment (poettering) [inline src/shared/generator.c:192]: i'd prefer if this part is split out into a separate commit within this PR

File: src/core/job.c [hunk 5]
```diff
@@ -949,22 +950,43 @@ static int job_dispatch_timer(sd_event_source *s, uint64_t monotonic, void *user
         return 0;
 }
 
-int job_start_timer(Job *j) {
+
```
Comment (poettering) [inline src/core/job.c]: spurious newline...

File: src/core/job.c [hunk 6]
```diff
@@ -949,22 +950,43 @@ static int job_dispatch_timer(sd_event_source *s, uint64_t monotonic, void *user
...
+                run_begin = now(CLOCK_MONOTONIC);
 
-        if (j->unit->job_timeout == USEC_INFINITY)
-                return 0;
+                if (j->timer_event_source) {
+                        /* Update only if JobRunningTimeout results in earlier timeout */
+                        (void) sd_event_source_get_time(j->timer_event_source, &timeout_time);
+                        if (timeout_time <= run_begin + j->unit->job_running_timeout)
+                                return 0;
+                        r = sd_event_source_set_time(j->timer_event_source, run_begin + j->unit->job_running_timeout);
```
Comment (poettering) [inline src/core/job.c]: please use usec_add() for adding timeouts, it deals somewhat sanely with overflows

File: src/core/unit.h [hunk 7]
```diff
@@ -114,6 +114,7 @@ struct Unit {
 
         /* Job timeout and action to take */
         usec_t job_timeout;
+        usec_t job_running_timeout;
```
Comment (poettering) [inline src/core/unit.h:117]: hmm, i don't really follow why you don't need two timer events here, for the two different timeouts? can you elaborate?

File: src/shared/generator.c (hunk 4 above)
Comment (Werkov) [inline src/shared/generator.c]: The two (finite) timeouts will always have a common end. JobTimeoutSec= is
spanning the JobRunningTimeoutSec= so there is only one timer event created for
the timeout that goes off earlier. (I find two events couter-productive if
equal/close values were configured for both.)

File: src/core/unit.h (hunk 7 above)
Comment (Werkov) [reply src/core/unit.h:117]: The two (finite) timeouts will always have a common end. JobTimeoutSec= is
spanning the JobRunningTimeoutSec= so there is only one timer event created for
the timeout that goes off earlier. (I find two events couter-productive if
equal/close values were configured for both.)

**Comment by Werkov** [comment]
Thanks for review. I force pushed the changes + commented to the question.

**Comment by Werkov** [comment]
Hi, I'm not sure if previous changes didn't fall through the cracks or you were just busy elsewhere -- so I send this message.

File: src/core/job.c [hunk 8]
```diff
@@ -949,22 +950,44 @@ static int job_dispatch_timer(sd_event_source *s, uint64_t monotonic, void *user
...
-                return 0;
+                if (j->timer_event_source) {
+                        /* Update only if JobRunningTimeoutSec= results in earlier timeout */
+                        (void) sd_event_source_get_time(j->timer_event_source, &old_timeout_time);
+
+                        if (old_timeout_time <= timeout_time)
+                                return 0;
+
+                        r = sd_event_source_set_time(j->timer_event_source, timeout_time);
+                        return r;
```
Comment (keszybz) [inline src/core/job.c]:
```C
return sd_event_source_set_time(...);
```

File: src/core/job.c [hunk 9]
```diff
@@ -949,22 +950,44 @@ static int job_dispatch_timer(sd_event_source *s, uint64_t monotonic, void *user
...
 
-        j->begin_usec = now(CLOCK_MONOTONIC);
+                run_begin = now(CLOCK_MONOTONIC);
+                timeout_time = usec_add(run_begin, j->unit->job_running_timeout);
 
-        if (j->unit->job_timeout == USEC_INFINITY)
-                return 0;
+                if (j->timer_event_source) {
+                        /* Update only if JobRunningTimeoutSec= results in earlier timeout */
+                        (void) sd_event_source_get_time(j->timer_event_source, &old_timeout_time);
```
Comment (keszybz) [inline src/core/job.c]: What happens if this calls returns an error? `old_timeout_time` will be used uninitialized.

File: src/core/unit.c [hunk 10]
```diff
@@ -1336,6 +1337,9 @@ int unit_load(Unit *u) {
                         goto fail;
                 }
 
+                if (u->job_running_timeout != USEC_INFINITY && u->job_running_timeout > u->job_timeout)
+                        log_unit_warning(u, "JobRunningTimeoutSec= is greater than JobTimeoutSec=, the latter will override.");
```
Comment (keszybz) [inline src/core/unit.c]: I don't like "override", because nothing is overridden. What about:
"JobRunningTimeoutSec= is greater than JobTimeoutSec=, it has no effect."

File: src/core/job.c (hunk 9 above)
Comment (Werkov) [reply src/core/job.c]: Currently `sd_event_source_get_time` always succeeds but you're right it's safer to fully honor its signature.

**Comment by Werkov** [comment]
Thanks @keszybz, I updated the branch addressing your remarks.

**Comment by keszybz** [comment]
Looks good, but there's a rebase error: in the first commit you change JobTimeoutSec→JobRunningTimeoutSec, then the other way in the second commit, and then back again in the third, in generator.c.

File: src/shared/generator.c [hunk 11]
```diff
@@ -188,10 +188,50 @@ int generator_write_timeouts(
...
+                return 0;
+
+        node = fstab_node_to_udev_node(what);
+        if (!node)
+                return log_oom();
+
+        if (!is_device_path(node)) {
+                /* Nothing to apply dependencies to. */
+                return 0;
+        }
```
Comment (poettering) [inline src/shared/generator.c]: nitpick, please drop {} around single line if blocks, see CODING_STYLE

File: src/shared/generator.c [hunk 12]
```diff
@@ -188,10 +188,50 @@ int generator_write_timeouts(
...
+                return log_error_errno(r, "Failed to make unit name from path: %m");
+
+        /* See mount_add_default_dependencies for explanation why we create such
+         * dependencies. */
+        return write_drop_in_format(dir, unit, 50, "netdev-dependencies",
+                                    "# Automatically generated by %1$s\n\n"
+                                    "[Unit]\nAfter=%2$s %3$s\nWants=%2$s",
+                                    program_invocation_short_name,
+                                    SPECIAL_NETWORK_ONLINE_TARGET,
+                                    SPECIAL_NETWORK_TARGET);
```
Comment (poettering) [inline src/shared/generator.c]: nitpick: Note that SPECIAL_NETWORK_TARGET and SPECIAL_NETWORK_ONLINE_TARGET are macros resolving to literal strings. Hence I think we should write this as:

```c
        return write_drop_in_format(dir, unit, 50, "netdev-dependencies",
                                    "# Automatically generated by %1$s\n\n"
                                    "[Unit]\n"
                                    "After=" SPECIAL_NETWORK_ONLINE_TARGET " " SPECIAL_NETWORK_TARGET"\n"
                                    "Wants=" SPECIAL_NETWORK_ONLINE_TARGET "\n",
                                    program_invocation_short_name);
```

This would have the benefit that we'd format this string at compile time, instead of runtime.

File: src/core/job.c [hunk 13]
```diff
@@ -949,22 +950,45 @@ static int job_dispatch_timer(sd_event_source *s, uint64_t monotonic, void *user
...
+                run_begin = now(CLOCK_MONOTONIC);
+                timeout_time = usec_add(run_begin, j->unit->job_running_timeout);
 
-        if (j->unit->job_timeout == USEC_INFINITY)
-                return 0;
+                if (j->timer_event_source) {
+                        /* Update only if JobRunningTimeoutSec= results in earlier timeout */
+                        r = sd_event_source_get_time(j->timer_event_source, &old_timeout_time);
+                        if (r < 0)
+                                return 0;
```
Comment (poettering) [inline src/core/job.c]: why are you eating up the error here?

File: src/shared/generator.c (hunk 12 above)
Comment (Werkov) [reply src/shared/generator.c]: Good point, thanks.

File: src/core/job.c (hunk 13 above)
Comment (Werkov) [reply src/core/job.c]: Argh, of course error should propagate, fixed.

**Comment by Werkov** [comment]
Thanks both. I force-pushed changes as per latest review comments.
